iam-token-manager/v1.ts

```diff
@@ -0,0 +1,190 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import extend = require('extend');
+import { sendRequest } from '../lib/requestwrapper';
+
+export type Options = {
+  iamApikey?: string;
+  iamAccessToken?: string;
+  iamUrl?: string;
+}
+
+export interface IamTokenData {
+  access_token: string;
+  refresh_token: string;
+  token_type: string;
+  expires_in: number;
+  expiration: number;
+}
+
+export class IamTokenManagerV1 {
+  name: string;
+  serviceVersion: string;
+  protected url: string;
+  protected tokenInfo: IamTokenData;
+  private iamApikey: string;
+  private userAccessToken: string;
+
+  /**
+   * IAM Token Manager Service
+   *
+   * Retreives, stores, and refreshes IAM tokens.
+   *
+   * @param {Object} options
+   * @param {String} options.iamApikey
+   * @param {String} options.iamAccessToken
+   * @param {String} options.iamUrl - url of the iam api to retrieve tokens from
+   * @constructor
+   */
+  constructor(options: Options) {
+    this.url = options.iamUrl || 'https://iam.ng.bluemix.net/identity/token';
+    this.tokenInfo = {} as IamTokenData;
+    if (options.iamApikey) {
+      this.iamApikey = options.iamApikey;
+    }
+    if (options.iamAccessToken) {
+      this.userAccessToken = options.iamAccessToken;
+    }
+  }
+
+  /**
+   * This function sends an access token back through a callback. The source of the token
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. If this class is managing tokens and does not yet have one, make a request for one
+   * 3. If this class is managing tokens and the token has expired, refresh it
+   * 4. If this class is managing tokens and has a valid token stored, send it
+   *
+   * @param {Function} cb - callback function that the token will be passed to
+   */
+  public getToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(null, this.userAccessToken);
+    } else if (!this.tokenInfo.access_token) {
+      // 2. request an initial token
+      this.requestToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else if (this.isTokenExpired()) {
+      // 3. refresh a token
+      this.refreshToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(null, this.tokenInfo.access_token);
+    }
+  }
```

Comment:
I like the helper functions that you used here. I think they make it very clear what's going on, even though this logic is relatively complex with multiple branches.

In particular, the use of `isTokenExpired`, `requestToken`, `refreshToken`, and `saveTokenInfo` is on point. 👌